### PR TITLE
Indentation clang-format 11

### PR DESCRIPTION
**Pull request by masterleinad**
Related to #11106, this shows the indentation with clang-format 11.
I am not opposed to updating but this will be a little annoying for everyone submitting a pull request at first. If we decide to do this, we also should also provide pre-compiled versions for various platforms which was a little of a struggle last time.

Some of the diffs are
```diff
diff --git a/include/deal.II/base/numbers.h b/include/deal.II/base/numbers.h
index 4e3432c7bc..452628021a 100644
--- a/include/deal.II/base/numbers.h
+++ b/include/deal.II/base/numbers.h
@@ -561,7 +561,7 @@ namespace numbers
 
   template <typename number>
   constexpr DEAL_II_CUDA_HOST_DEV const number &
-                                        NumberTraits<number>::conjugate(const number &x)
+  NumberTraits<number>::conjugate(const number &x)
   {
     return x;
   }
@@ -690,7 +690,7 @@ namespace internal
   struct NumberType
   {
     static constexpr DEAL_II_ALWAYS_INLINE DEAL_II_CUDA_HOST_DEV const T &
-                                                                       value(const T &t)
+    value(const T &t)
     {
       return t;
     }
@@ -705,8 +705,8 @@ namespace internal
     // Type T is constructible from F.
     template <typename F>
     static constexpr DEAL_II_ALWAYS_INLINE DEAL_II_CUDA_HOST_DEV T
-                                                                 value(const F &f,
-                                                                       typename std::enable_if<
+    value(const F &f,
+          typename std::enable_if<
             !std::is_same<typename std::decay<T>::type,
                           typename std::decay<F>::type>::value &&
             std::is_constructible<T, F>::value>::type * = nullptr)
@@ -717,8 +717,8 @@ namespace internal
     // Type T is explicitly convertible (but not constructible) from F.
     template <typename F>
     static constexpr DEAL_II_ALWAYS_INLINE T
-                                           value(const F &f,
-                                                 typename std::enable_if<
+    value(const F &f,
+          typename std::enable_if<
             !std::is_same<typename std::decay<T>::type,
                           typename std::decay<F>::type>::value &&
             !std::is_constructible<T, F>::value &&
```
so there are at least some improvements.

### CODE REVIEW

File: examples/step-50/step-50.cc
Comment (tjhei) [inline examples/step-50/step-50.cc:507]: ugly

File: examples/step-21/step-21.cc
Comment (tjhei) [inline examples/step-21/step-21.cc:1177]: I don't like this behavior.

**Review by tjhei — COMMENTED** [review]
I am not sure if switching makes sense or not, but I dislike some of the changes. I assume some of them can be controlled by options.

**Comment by tjhei** [comment]
> I assume some of them can be controlled by options.

Before you spend more time on this, @masterleinad , we should decide what to do.

Does the switch fix the problems in #11106?

**Comment by kronbichler** [comment]
> Does the switch fix the problems in #11106?

At least the one in #8437 looks much better from what I see in the top post. If that alone is enough the pain with a different version, I don't know. I am nonetheless positive about it, albeit not very strongly and only if we have a majority of maintainers positive to this topic.

File: examples/step-28/step-28.cc
Comment (bangerth) [inline examples/step-28/step-28.cc:1666]: This seems wrong -- the closing brace no longer lines up vertically with the opening brace...

File: examples/step-33/step-33.cc
Comment (bangerth) [inline examples/step-33/step-33.cc:2125]: I also don't understand this one -- it's indenting the comment four (instead of two) characters relative to the opening brace.

File: examples/step-41/step-41.cc
Comment (bangerth) [inline examples/step-41/step-41.cc:88]: This, at least, is progress :-)

**Comment by tamiko** [comment]
This will also require to update our download and compile magic for `clang-format`. But first things first, we should find a consensus whether `clang-format-11` is finally a version we want to update to.

**Comment by tamiko** [comment]
@masterleinad Ah, now I understand. Would you mind to update `examples/.clang-format` as well? :smile:

File: include/deal.II/integrators/patches.h
Comment (tamiko) [inline include/deal.II/integrators/patches.h:43]: I am a bit dissapointed that clang-format still cannot align the `&` properly :-(

**Comment by tamiko** [comment]
I think that this configuration with clang-format-11 is now a strict improvement over what we had before and doesn't introduce any new unwanted behavior. Let us discuss at some point whether this change is worth the infrastructure update. :smile:

File: include/deal.II/base/table_indices.h
Comment (drwells) [inline include/deal.II/base/table_indices.h:70]: These look weird but now that we have C++14 perhaps we can fix it by just promoting these from `DEAL_II_CONSTEXPR` to `constexpr`.

Comment (masterleinad) [reply include/deal.II/base/table_indices.h:70]: We know that at least MSVC has problems and the promotion would not work in general. https://github.com/dealii/dealii/blob/master/cmake/checks/check_01_cxx_features.cmake#L572-L621
Of course, we could try in selected places.

Comment (drwells) [reply include/deal.II/base/table_indices.h:70]: I'll give it a try

**Comment by bangerth** [comment]
Here's another consistent source of poor indentation: rigidly sticking to 80 chars per line. This forces the tool to break lines, even if they are just a bit longer, and there are many places where we get poor line breaks, or where stuff isn't vertically aligned because it doesn't fit. For example (`step-22.cc`):

* Here, we could have nicely aligned everything if we had more space
```
                symgrad_phi_u[k] =
                  fe_values[velocities].symmetric_gradient(k, q);
                div_phi_u[k] = fe_values[velocities].divergence(k, q);
                phi_p[k]     = fe_values[pressure].value(k, q);
```
or here, where the second argument to the constructor is not aligned with the first one:
```
      SolverControl            solver_control(solution.block(1).size(),
                                   1e-6 * schur_rhs.l2_norm());
```
or here, where nothing is really aligned:
```
  Assert(
    this->at_boundary(),
    ExcMessage(
      "You are trying to set the boundary_id of an internal object, which is illegal!"));
```

I think that a lot of these issues would go away if we gave ourselves 90 or 100 characters.  I think that also shouldn't be an issue any more. The 80 character convention originates in how many characters fit onto a text-only screen, and matches what fit onto a punch card. But our screens are wider and fit more text today -- for example, the longest lines you see in the following picture are 77 characters wide:
image
There is certainly a limit to what is easily readable, but I don't see a good reason not to go for 100 characters -- in practice, quite few lines will actually be this long.

**Comment by kronbichler** [comment]
I would be very happy with a 100 character limit of lines (but we really need a quiet time for that because we would change a considerable number of lines).

**Comment by masterleinad** [comment]
I propose to look into this right after the release.

File: contrib/utilities/indent_common.sh
Comment (marcfehling) [inline contrib/utilities/indent_common.sh:61]:
```suggestion
    echo "***   This indent script requires clang-format version 11.0,"
```

**Comment by masterleinad** [comment]
We need to upload binaries for `Linux` and `MacOS` but otherwise, this should be good already.

**Comment by tamiko** [comment]
@masterleinad 75% done, just waiting for the mac binary(ies). I suggest once ready, you simply incorporate #12445 into this PR?

**Comment by masterleinad** [comment]
> @masterleinad 75% done, just waiting for the mac binary(ies). I suggest once ready, you simply incorporate #12445 into this PR?

Yes, already done. The indentation check is passing, too.

File: include/deal.II/base/table_indices.h
Comment (drwells) [reply include/deal.II/base/table_indices.h]: We fixed a lot of these in #11122 but we can't yet get rid of `DEAL_II_CONSTEXPR` - we need it to compile with some older compilers.

**Comment by masterleinad** [comment]
It looks like only updating @tjhei's CI scripts is missing here now. Since these containers don't get updated automatically, this needs to be done at the same time as merging this pull request.

**Comment by masterleinad** [comment]
Updating the clang-format version in the CI is also necessary to run it at all.

**Review by tamiko — APPROVED** [review]

**Review by kronbichler — APPROVED** [review]

**Comment by luca-heltai** [comment]
> Updating the clang-format version in the CI is also necessary to run it at all.

This is done automatically in the actions:

```
 - name: setup
      run: |
        ./contrib/utilities/download_clang_format
        sudo apt update && sudo apt install doxygen graphviz perl texlive-bibtex-extra
        doxygen --version
```

**Comment by masterleinad** [comment]
@luca-heltai What I meant is that @tjhei's containers don't update automatically (and that's why one of the `indent` steps and the `CI` step is failing).

**Comment by luca-heltai** [comment]
Sorry, I thought we agreed that we should remove the indent step from the CI.

**Review by luca-heltai — APPROVED** [review]

**Review by kronbichler — APPROVED** [review]

**Comment by masterleinad** [comment]
> Sorry, I thought we agreed that we should remove the indent step from Timo's CI.

I think we agreed on removing the `indent` check but I'm not sure about the `CI` step. I see some value in only trigger these expensive tests when at least the indentation is correct.

**Comment by masterleinad** [comment]
@tjhei Any chance that we can move forward here (disabling the indentation check for the CI on your side or upgrading it at the time as merging this pull request)?

**Comment by tjhei** [comment]
sure, you are free to disable it for now.

**Comment by masterleinad** [comment]
> sure, you are free to disable it for now.

AFAICT, neither `CI` nor `indent` is defined within the repository; I only see Jenkinsfiles for `ready`, `OSX`, and `tidy`. Someone with `owner` permission status can disable `tidy` (since it is redundant due to a corresponding `GitHub` action) but it seems that for `CI` some action by @tjhei is still required.

**Comment by tjhei** [comment]
Did you see ./Jenkinsfile ?

**Comment by masterleinad** [comment]
> Did you see ./Jenkinsfile ?

Seems that I missed that. Thanks!

**Review by tamiko — APPROVED** [review]

**Review by kronbichler — APPROVED** [review]

**Comment by tjhei** [comment]
> disable tidy (since it is redundant 

Wait, where are we running clang tidy already?

**Comment by masterleinad** [comment]
> Wait, where are we running clang tidy already?

Oh, typo. I meant `format`. As can be seen in the checks `tidy` is still running.

**Review by peterrum — APPROVED** [review]

**Review by marcfehling — APPROVED** [review]

**Pull request merged**
